crates/dekaf/src/log_journal.rs

```diff
@@ -0,0 +1,344 @@
+                journal: journal.to_owned(),
+                ..Default::default()
+            },
+            || {
+                futures::stream::once({
+                    let value = log_data.clone();
+                    async move { Ok(value) }
+                })
+            },
+        );
+
+        tokio::pin!(resp);
+
+        let mut attempts = 0;
+
+        loop {
+            match resp.try_next().await {
+                Err(err) if err.is_transient() && attempts < 3 => {
+                    attempts += 1;
+                }
+                Err(err) => {
+                    if err.is_transient() {
+                        tracing::warn!(
+                            ?err,
+                            attempts,
+                            "Got recoverable error multiple times while trying to write logs"
+                        )
+                    }
+                    return Err(err.into());
+                }
+                Ok(_) => return Ok(()),
+            }
+        }
+    }
+
+    fn serialize_log(&self, log: ops::Log) -> Vec<u8> {
+        let uuid = gazette::uuid::build(
+            self.producer,
+            gazette::uuid::Clock::from_time(std::time::SystemTime::now()),
+            uuid::Flags(
+                (message_flags::MASK | message_flags::OUTSIDE_TXN)
+                    .try_into()
+                    .unwrap(),
+            ),
+        );
+
+        let mut val = serde_json::to_value(log).expect("Log always serializes");
+
+        if let Some(obj) = val.as_object_mut() {
+            obj.insert("_meta".to_string(), json!({ "uuid": uuid }));
+        }
+
+        let mut buf = serde_json::to_vec(&val).expect("Value always serializes");
+        buf.push(b'\n');
+
+        buf
+    }
+
+    async fn get_journal_client(
+        &self,
+        task_name: String,
+    ) -> anyhow::Result<(journal::Client, String)> {
+        let (client, _claims, ops_logs, _ops_stats, _task_spec) = fetch_dekaf_task_auth(
+            self.app.client_base.clone(),
+            &task_name,
+            &self.app.data_plane_fqdn,
+            &self.app.data_plane_signer,
+        )
+        .await?;
+
+        let client = fetch_task_authorization(
+            &client,
+            &dekaf_shard_template_id(task_name.as_str()),
+            &self.app.data_plane_fqdn,
+            &self.app.data_plane_signer,
+            proto_flow::capability::AUTHORIZE | proto_gazette::capability::APPEND,
+            gazette::broker::LabelSelector {
+                include: Some(labels::build_set([("name", ops_logs.as_str())])),
+                exclude: None,
+            },
+        )
+        .await?;
+
+        Ok((client, ops_logs))
+    }
+}
+
+#[derive(Clone)]
+struct SessionLogger {
+    tx: tokio::sync::mpsc::Sender<LoggingMessage>,
+    _handle: Arc<JoinHandle<()>>,
+}
+
+impl SessionLogger {
+    fn new(app: Arc<App>, producer: Producer) -> Self {
+        let (log_tx, log_rx) = tokio::sync::mpsc::channel(1000);
```

Comment:
This is probably not ideal.. I'd like to refactor so that setting the task name is a oneshot and this is more like a stack that only contains the N newest items, rather than erroring once this channel gets full.